design-proposals/klip-49-add-source-stream-table-semantic.md

```diff
@@ -0,0 +1,161 @@
+could be potentially terminated directly by the end user, which comes with a consistency problem with source table 
+semantic agreement. If a source table query could be terminated independently by the user, the original source table 
+would be dangling and not available for pull query anymore. To address this problem, we have two approaches:
+ 
+ 1. The KSQL server rejects the termination of a source table query and informs end user to try deleting the original source table.
+ 2. The KSQL server will verify the deletion of a query, and try deleting the associated source table with it.
+ 
+Both approaches are valid here to maintain the consistency, but considering when dropping a source table, the active 
+materialization query would be terminated, I'm inclined to suggest taking approach #1 here for simplicity, and avoid 
+introducing additional code path to have direct metadata affection when we could just make it as a one-way deletion for 
+source table.  
+
+For the materialized store in source table, there would also be a backing-up changelog topic. The reasoning is that the input 
+topic could potentially be using a different serde than the internal serde being used by KSQL. Having asymmetric serde for 
+serialization and deserialization could lead to potentially state store pollution as we do bulk-loading of source topic 
+data into the state store. The similar [issue](https://github.com/confluentinc/ksql/issues/5673) has been raised before where 
+KSQL has to turn off topology optimization to avoid having state stores backed up by non-internal topics for changelog.   
+
+#### Pull Query Support
+One of the motivation for table materialization is to support pull query, but it comes with a need for having the underlying 
+table partitioned by the primary key, in order to ensure pull query works. Here the debate was that whether we should enforce 
+a repartition of the input topic blindly to guarantee that it is using the primary key as expected for sharding. The current plan 
+would be not worrying about the undefined user behavior during the source table primary key construction here, but instead 
+just assume they gave us a well-partitioned input topic whose materialization is ready for pull query. This aligns with 
+our original expectation of a SOURCE TABLE as well, where KSQL should not have extra mutation or tweak on the original 
+data unless specified via CTAS or other query.  
+
+## Public APIS
+Create source table/stream, which makes the data source read-only:
+
+```roomsql
+CREATE SOURCE STREAM stream_name (COL1 INT, COL2 STRING) WITH ...
+CREATE SOURCE TABLE table_name (ID INT PRIMARY KEY, COL1 INT) WITH ...
+```
+
+## Test plan
+The new semantics will be tested as:
+
+* Unit testing for modular changes
+* QTTs with the new source table/stream syntax, including insertion failure, `DROP ... DELETE TOPIC` command integrity
```

Comment:
QTTs don't support `DROP STREAM/TABLE`. We'll have to write separate integration tests (separate from the QTT framework for this). 

(QTTs also don't support INSERT VALUES but RQTT does so we're fine on that one.)

Reply:
There's a drop_source.json file and an insert-values.json file in the test folder that contain the commands, so we should be fine with QTTs

Reply:
`insert-values.json` is the RQTT file I was referring to, yeah.

I didn't know about `drop_source.json`. I don't think `DROP ... DELETE TOPIC;` is going to work, though. (The fact that QTT is being used to test `DROP STREAM/TABLE` statements at all is sort of a hack since that's not what it's meant to do. The tests in `drop_source.json` have dummy CSAS statements for this reason.)